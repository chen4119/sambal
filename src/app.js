import {LitElement} from '@polymer/lit-element';
import {installRouter} from 'pwa-helpers/router.js';
import {installMediaQueryWatcher} from 'pwa-helpers/media-query.js';
import {store} from './store.js';
import {updateLocation, updateScreenSize} from './actions/sambal.js';

export class SambalApp extends LitElement {

    constructor(routes, smallScreenWidth = 767) {
        super();
        this.routeMap = new Map();
        for (let i = 0; i < routes.length; i++) {
            this.routeMap.set(routes[i].path, routes[i]);
        }
        this.smallScreenWidth = smallScreenWidth;
    }

    firstUpdated() {
        installRouter((location) => {
            const path = decodeURIComponent(location.pathname);
            store.dispatch(updateLocation(path));
            if (this.routeMap.has(path)) {
                const route = this.routeMap.get(path);
                if (route.dispatch) {
                    store.dispatch(route.dispatch);
                }
            }
        });
        installMediaQueryWatcher(`(max-width: ${this.smallScreenWidth}px)`, (matches) => store.dispatch(updateScreenSize(matches)));

        // Custom elements polyfill safe way to indicate an element has been upgraded.
        this.removeAttribute('unresolved');
    }

    getRoute(path) {
        return this.routeMap.get(path);
    }

}