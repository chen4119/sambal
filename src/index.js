export {SambalApp} from "./app";
export {store} from "./store";