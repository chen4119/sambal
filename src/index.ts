import LinkedDataStore from "./LinkedDataStore";
import Packager from "./Packager";

export {
    toJsonLdGraph,
    toSchemaOrgJsonLd,
    graphToCondensedJsonLd,
    hydrateJsonLd
} from "sambal-jsonld";

export {LinkedDataStore};
export {Packager};
export {template} from "./template";
export {render} from "./operators/render";
export {aggregateBy} from "./operators/aggregateBy";
export {paginate} from "./operators/paginate";
export {toHtml} from "./operators/toHtml";
export {pushSchemaOrgJsonLd} from "./operators/pushSchemaOrgJsonLd";
export {OUTPUT_FOLDER, CACHE_FOLDER} from "./constants";