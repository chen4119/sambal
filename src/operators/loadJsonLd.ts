import {Observable, pipe} from "rxjs";
import {mergeMap} from "rxjs/operators";
import {loadContent} from "../utils";
import {hydrateJsonLd} from "sambal-jsonld";

export function loadJsonLd(hydrateOptions?: {
    fetcher?: (url: string) => Promise<any>,
    context?: any
}) {
    return pipe<Observable<string>, Observable<any>>(
        mergeMap(async src => {
            const content = await loadContent(src);
            if (hydrateOptions) {
                return await hydrateJsonLd(content, hydrateOptions.fetcher, hydrateOptions.context);
            }
            return content;
        })
    );
}