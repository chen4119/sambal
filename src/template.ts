
export async function template(literals, ...expressions) {
    const resolvedExpr = await Promise.all(expressions);
    let string = "";
    for (const [i, val] of resolvedExpr.entries()) {
        if (Array.isArray(val)) {
            const resolvedValues = await Promise.all(val);
            string += `${literals[i]}${resolvedValues.filter(v => v !== null && v !== undefined).join("")}`;
        } else {
            string += `${literals[i]}${(val === null || val === undefined) ? "" : val}`;
        }
    }
    string += literals[literals.length - 1];
    return string;
}
