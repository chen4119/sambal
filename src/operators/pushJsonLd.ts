import {pipe, Observable} from "rxjs";
import {map} from "rxjs/operators";
import {SambalData, SAMBAL_INTERNAL} from "../constants";

export function pushJsonLd(transformer: (props: any) => any) {
    return pipe<Observable<any>, Observable<SambalData>>(
        map((data) => {
            if (!data[SAMBAL_INTERNAL]) {
                data[SAMBAL_INTERNAL] = {};
            }
            if (!data[SAMBAL_INTERNAL].jsonld) {
                data[SAMBAL_INTERNAL].jsonld = [];
            }
            const jsonld = transformer(data);
            if (jsonld) {
                data[SAMBAL_INTERNAL].jsonld.push(jsonld);
            }
            return data;
        })
    );
}
