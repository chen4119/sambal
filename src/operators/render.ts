import {pipe, Observable} from "rxjs";
import {mergeMap} from "rxjs/operators";
import * as cheerio from "cheerio";
import LocalCss from "../LocalCss";
import {SambalData, SAMBAL_INTERNAL} from "../constants";

export function render(renderer: (props: any) => Promise<string>) {
    return pipe<Observable<SambalData>, Observable<SambalData>>(
        mergeMap(async (data: SambalData) => {
            const css = new LocalCss();
            const html = await renderer({...data, css: css});
            const $ = cheerio.load(html);
            if (css.hasSheets()) {
                $("head").append(`
                    <style>
                        ${css.getCss()}
                    </style>
                `);
            }
            if (!data[SAMBAL_INTERNAL]) {
                data[SAMBAL_INTERNAL] = {uri: ""};
            }
            data[SAMBAL_INTERNAL].html = $;
            return data;
        })
    );
}
