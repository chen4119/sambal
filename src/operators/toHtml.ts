import {pipe} from "rxjs";
import {map} from "rxjs/operators";

export function toHtml() {
    return pipe(
        map<{html: CheerioStatic}, string>((d) => {
            const $ = d.html;
            return $.html();
        })
    );
}