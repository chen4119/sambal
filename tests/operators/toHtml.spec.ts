import {of, from} from "rxjs";
import {template} from "../../src/template";
import {loadJsonLd} from "../../src/operators/loadJsonLd";
import {render} from "../../src/operators/render";
import {toHtml} from "../../src/operators/toHtml";
import {pushJsonLd} from "../../src/operators/pushJsonLd";
import {toSchemaOrgJsonLd} from "sambal-jsonld";

describe('toHtml', () => {
    it('add css to head', async () => {
        const result = await of({text: "hello world"})
        .pipe(render(({css, text}) => {
            const classes = css.style({
                main: {
                    "font-style": "italic"
                }
            });
            return template`
                <!doctype html>
                <html>
                    <head>
                        <meta charset="UTF-8">
                    </head>
                    <body>
                        ${template`
                            <main class=${classes.main}>
                                ${text}
                            </main>
                        `}
                    </body>
                </html>
            `;
        }))
        .pipe(toHtml())
        .toPromise();
        expect(result).toMatchSnapshot();
    });

    it('add schema.org to head', async () => {
        const result = await from(['tests/post.md'])
        .pipe(loadJsonLd())
        .pipe(pushJsonLd(d => toSchemaOrgJsonLd(d, "BlogPosting")))
        .pipe(render(({text}) => {
            return template`
                <!doctype html>
                <html>
                    <head>
                        <meta charset="UTF-8">
                    </head>
                    <body>
                        ${template`
                            <main>
                                ${text}
                            </main>
                        `}
                    </body>
                </html>
            `;
        }))
        .pipe(toHtml())
        .toPromise();
        expect(result).toMatchSnapshot();
    });

    it('automatically add head if not found', async () => {
        const result = await of({text: "hello world"})
        .pipe(render(({css, text}) => {
            const classes = css.style({
                main: {
                    "font-style": "italic"
                }
            });
            return template`
                <!doctype html>
                <html>
                    <body>
                        ${template`
                            <main class=${classes.main}>
                                ${text}
                            </main>
                        `}
                    </body>
                </html>
            `;
        }))
        .pipe(toHtml())
        .toPromise();
        expect(result).toMatchSnapshot();
    });

    it('replace img', async () => {
        const result = await of({text: "hello world"})
        .pipe(render(({css, text}) => {
            const classes = css.style({
                main: {
                    "font-style": "italic"
                }
            });
            return template`
                <!doctype html>
                <html>
                    <body>
                        ${template`
                            <main class=${classes.main}>
                                ${text}
                            <img src="/img1.png">
                            </main>
                        `}
                    </body>
                </html>
            `;
        }))
        .pipe(toHtml({
            replace: (name, attribs) => {
                if (name === 'img') {
                    return {
                        name: 'picture',
                        body: [
                            {name: 'source', attributes: {media: "(min-width: 480px)", srcset: "img1.jpg"}},
                            {name: 'source', attributes: {media: "(min-width: 650px)", srcset: "img2.jpg"}}
                        ]
                    };
                }
                return null;
            }
        }))
        .toPromise();
        expect(result).toMatchSnapshot();
    });
})