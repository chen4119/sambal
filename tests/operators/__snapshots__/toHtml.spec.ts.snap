// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toHtml add css to head 1`] = `
"<!doctype html>
<html>
    <head>
        <meta charset=\\"UTF-8\\">
        <style>.main-0-0-1 {
  font-style: italic;
}</style>
    </head>
    <body>
        <main class=\\"main-0-0-1\\">
                                hello world
                            </main>
    </body>
</html>"
`;

exports[`toHtml add schema.org to head 1`] = `
"<!doctype html>
<html>
    <head>
        <meta charset=\\"UTF-8\\">
        <script type=\\"application/ld+json\\">{\\"@context\\":\\"https://schema.org\\",\\"@graph\\":[{\\"headline\\":\\"Second post\\",\\"description\\":\\"Some description\\",\\"keywords\\":[\\"javascript\\",\\"rxjs\\"],\\"dateCreated\\":\\"2019-03-03T19:00:00.000-05:00\\",\\"text\\":\\"<p>This is my very first post1</p>\\\\n<pre><code class=\\\\\\"language-ShellSession\\\\\\">$ npm install --save-dev sambal-cli</code></pre>\\\\n<p>Sambal has <a href=\\\\\\"https://webpack.org\\\\\\">webpack</a>, <a href=\\\\\\"https://webpack.org\\\\\\">webpack</a>, <a href=\\\\\\"https://webpack.org\\\\\\">webpack</a> to do with something else</p>\\\\n<pre><code class=\\\\\\"language-js\\\\\\">import {pipe, from, of} from &quot;rxjs&quot;;\\\\nimport {mergeMap, map, groupBy, reduce, toArray} from &quot;rxjs/operators&quot;;\\\\n\\\\nfunction aggregateBy(field) {\\\\n    return pipe(\\\\n        map((data) =&gt; data[field]),\\\\n        mergeMap(fieldValue =&gt; Array.isArray(fieldValue) ? from(fieldValue) : of(fieldValue)),\\\\n        groupBy(v =&gt; v),\\\\n        mergeMap(group$ =&gt;\\\\n            group$.pipe(reduce((acc, cur) =&gt; {\\\\n                acc.count++;\\\\n                return acc;\\\\n            }, {key: group$.key, count: 0}))    \\\\n        ),\\\\n        toArray()\\\\n    );\\\\n}</code></pre>\\\\n\\",\\"@id\\":\\"_:1\\",\\"@context\\":\\"https://schema.org\\",\\"@type\\":\\"BlogPosting\\"}]}</script>
    </head>
    <body>
        <main>
            <p>This is my very first post1</p>
            <pre><code class=\\"language-ShellSession\\">$ npm install --save-dev sambal-cli</code></pre>

            <p>Sambal has <a href=\\"https://webpack.org\\">webpack</a>, <a href=\\"https://webpack.org\\">webpack</a>, <a href=\\"https://webpack.org\\">webpack</a> to do with something else</p>
            <pre><code class=\\"language-js\\">import {pipe, from, of} from &quot;rxjs&quot;;
import {mergeMap, map, groupBy, reduce, toArray} from &quot;rxjs/operators&quot;;

function aggregateBy(field) {
    return pipe(
        map((data) =&gt; data[field]),
        mergeMap(fieldValue =&gt; Array.isArray(fieldValue) ? from(fieldValue) : of(fieldValue)),
        groupBy(v =&gt; v),
        mergeMap(group$ =&gt;
            group$.pipe(reduce((acc, cur) =&gt; {
                acc.count++;
                return acc;
            }, {key: group$.key, count: 0}))    
        ),
        toArray()
    );
}</code></pre>

        </main>
    </body>
</html>"
`;

exports[`toHtml automatically add head if not found 1`] = `
"<!doctype html>
<html>
    <head>
        <style>.main-0-0-2 {
  font-style: italic;
}</style>
    </head>
    <body>
        <main class=\\"main-0-0-2\\">
                                hello world
                            </main>
    </body>
</html>"
`;

exports[`toHtml replace img 1`] = `
"<!doctype html>
<html>
    <head>
        <style>.main-0-0-3 {
  font-style: italic;
}</style>
    </head>
    <body>
        <main class=\\"main-0-0-3\\">
                                hello world
                            <picture>
            <source media=\\"(min-width: 480px)\\" srcset=\\"img1.jpg\\">
            <source media=\\"(min-width: 650px)\\" srcset=\\"img2.jpg\\">
        </picture>
    </main>
</body>
</html>"
`;
